### ISSUE TITLE
Sheaves Discussion

### ISSUE THREAD

**Issue by simonbrandhorst** [HechtiDerLachs/Oscar.jl#18]

**Comment by simonbrandhorst** [comment]
https://github.com/HechtiDerLachs/Oscar.jl/blob/sheaf_interface/experimental/Schemes/Sheaves.jl#L63-L73
Shouldn't that rather be part of `X`?

**Comment by simonbrandhorst** [comment]
$\mathcal{F}(U,V)$ could be an alias for `restriction_map(F,U,V)`?

**Comment by HechtiDerLachs** [comment]
> https://github.com/HechtiDerLachs/Oscar.jl/blob/sheaf_interface/experimental/Schemes/Sheaves.jl#L63-L73 Shouldn't that rather be part of `X`?

Both. The idea is that the user has the possibility to provide a custom function. That could, for instance, take care of special caching in the background, or, say for technical reasons, single out only special types of inputs to be admissible, etc., etc. 

When you look at the corresponding default argument for the constructor, you see that this is already set to `is_open_embedding`, so it does refer to functionality of `X`, eventually.



> F(U,V) could be an alias for `restriction_map(F,U,V)`?

It already is.

**Comment by HechtiDerLachs** [comment]
I just did a sample implementation for the structure sheaf. Try `SheafOO(X)` for any `AbsSpec` `X` of your choice.

**Comment by HechtiDerLachs** [comment]
`SheafOO` now also works for covered schemes. See the test file.

**Comment by afkafkafk13** [comment]
The names is_open_func and production_func are far from being self-explanatory.  They only started to make sense to me, when reading the structure sheaf example. Isn't production_func just something like sections_over_open? For the is_open_func I cannot come up with a good formulation, but you want to check whether we are in a situation where it makes sense to ask for the restriction map \rho_{V,U}, right?

**Comment by HechtiDerLachs** [comment]
> but you want to check whether we are in a situation where it makes sense to ask for the restriction map \rho_{V,U}, right?

Yes, exactly that. It is 'explained' [here](https://github.com/HechtiDerLachs/Oscar.jl/blob/sheaf_interface/experimental/Schemes/Sheaves.jl#L96). 

The point is: A sheaf is necessarily a lazy object; similar to automatically extended chain complexes. Therefore, we need to 'produce' an element on request and store it for future reference. Whether or not a certain input key is admissible, has to be decided and since the code is generic, we ask the user to provide us with a function `is_open_func` that takes care of this decision. Likewise, the user has to provide a function `production_func` and `restriction_func` to produce the value of the sheaf on an admissible open set and the restriction map for an inclusion of open sets. 

I know that this is not self explanatory. But keep in mind that this is merely a first prototype to demonstrate how such things could be implemented.

**Comment by afkafkafk13** [comment]
Yes, I had already guessed this from the code -- and this draft of the concept looks good to me.

**Comment by simonbrandhorst** [comment]
I am not sure if the parametrization by `RestrictionType` is really helpful/necessary.
Apart from this, the Sheaves look pretty solid. Good job.
However, for now our priority is to get what we have to a solid, well tested and well documented state. 
So please stop working on this matter until we are at 80-90% coverage for all functions in `Schemes` .
